File: zeroinstall/cmd/whatchanged.py
"""
The B{0install whatchanged} command-line interface.
"""

# Copyright (C) 2012, Thomas Leonard
# See the README file for details, or visit http://0install.net.

from __future__ import print_function

import os

from zeroinstall import _, SafeException
from zeroinstall.cmd import UsageError

syntax = "APP-NAME"

def add_options(parser):
	parser.add_option("", "--full", help=_("show diff of the XML"), action='store_true')

def handle(config, options, args):
	if len(args) != 1:
		raise UsageError()

	name = args[0]
	app = config.app_mgr.lookup_app(name, missing_ok = False)
	history = app.get_history()

	if not history:
		raise SafeException(_("Invalid application: no selections found! Try '0install destroy {name}'").format(name = name))

	import time

	last_checked = app.get_last_checked()
	if last_checked is not None:
		print(_("Last checked    : {date}").format(date = time.ctime(last_checked)))

	last_attempt = app.get_last_check_attempt()
	if last_attempt is not None:
		print(_("Last attempted update: {date}").format(date = time.ctime(last_attempt)))

	print(_("Last update     : {date}").format(date = history[0]))
	current_sels = app.get_selections(snapshot_date = history[0])

	if len(history) == 1:
		print(_("No previous history to compare against."))
		print(_("Use '0install select {name}' to see the current selections.").format(name = name))
		return

	print(_("Previous update : {date}").format(date = history[1]))

	def get_selections_path(date):
		return os.path.join(app.path, 'selections-{date}.xml'.format(date = date))

	print()

	if options.full:
		import difflib, sys
		def load_lines(date):
			with open(get_selections_path(date), 'r') as stream:
				return stream.readlines()
		old_lines = load_lines(history[1])
		new_lines = load_lines(history[0])
		for line in difflib.unified_diff(old_lines, new_lines, fromfile = history[1], tofile = history[0]):
			sys.stdout.write(line)
	else:
		changes = show_changes(app.get_selections(snapshot_date = history[1]).selections, current_sels.selections)
		if not changes:
			print(_("No changes to versions (use --full to see all changes)."))

	print()
	print(_("To run using the previous selections, use:"))
	print("0install run {path}".format(path = get_selections_path(history[1])))

def show_changes(old_selections, new_selections):
	changes = False

	for iface, old_sel in old_selections.iteritems():
		new_sel = new_selections.get(iface, None)
		if new_sel is None:
			print(_("No longer used: %s") % iface)
			changes = True
		elif old_sel.version != new_sel.version:
			print(_("%s: %s -> %s") % (iface, old_sel.version, new_sel.version))
			changes = True

	for iface, new_sel in new_selections.iteritems():
		if iface not in old_selections:
			print(_("%s: new -> %s") % (iface, new_sel.version))
			changes = True
	
	return changes

File: ocaml/whatchanged.ml
(* Copyright (C) 2013, Thomas Leonard
 * See the README file for details, or visit http://0install.net.
 *)

(** The "0install whatchanged" command *)

open Options
open Zeroinstall.General
open Support.Common
module Qdom = Support.Qdom
module U = Support.Utils
module Selections = Zeroinstall.Selections

let show_changes (system:system) old_selections new_selections =
  let changes = ref false in
  let v sel = Zeroinstall.Element.version sel in
  let print fmt = Support.Utils.print system fmt in

  old_selections
  |> if_some (Selections.iter (fun iface old_sel ->
    match Selections.find iface new_selections with
    | None ->
        print "No longer used: %s" iface;
        changes := true
    | Some new_sel ->
        if (v old_sel) <> (v new_sel) then (
          print "%s: %s -> %s" iface (v old_sel) (v new_sel);
          changes := true
        )
  ));

  new_selections |> Selections.iter (fun iface new_sel ->
    let old_sel = old_selections |> pipe_some (Selections.find iface) in
    if old_sel = None then (
      print "%s: new -> %s" iface (v new_sel);
      changes := true
    )
  );
  
  !changes

let show_app_changes options ~full app =
  let module A = Zeroinstall.Apps in
  let config = options.config in
  let system = config.system in
  let print fmt = U.print system fmt in

  match A.get_history config app with
  | [] -> raise_safe "Invalid application: no selections found! Try '0install destroy %s'" (Filename.basename app)
  | current :: history ->
      let times = A.get_times system app in
      if times.A.last_check_time <> 0.0 then
        print "Last checked    : %s" (U.format_time_pretty @@ Unix.gmtime times.A.last_check_time);
      let () = 
        match times.A.last_check_attempt with
        | Some attempt ->
            print "Last attempt    : %s" (U.format_time_pretty @@ Unix.gmtime attempt);
        | None -> () in

      print "Last update     : %s" current;
      let current_sels = A.get_selections_no_updates system app in

      match history with
      | [] ->
          print "No previous history to compare against.";
          print "Use \"0install show %s\" to see the current selections." (Filename.basename app);
      | previous :: _ ->
          print "Previous update : %s" previous;

          let get_selections_path date = app +/ Printf.sprintf "selections-%s.xml" date in

          print "";

          if full then (
            let argv = ["diff"; "-u"; "--"; get_selections_path previous; get_selections_path current] in
            let child = system#create_process argv Unix.stdin Unix.stdout Unix.stderr in
            flush stdout;
            match snd @@ system#waitpid_non_intr child with
            | Unix.WEXITED (0|1) -> ()    (* OK, 1 just means changes found *)
            | x -> Support.System.check_exit_status x
          ) else (
            let old_sels = Zeroinstall.Selections.load_selections system @@ get_selections_path previous in
            let changes = show_changes system (Some old_sels) current_sels in
            if not changes then
              print "No changes to versions (use --full to see all changes)."
          );

          print "";
          print "To run using the previous selections, use:";
          print "0install run %s" (get_selections_path previous)

let handle options flags args =
  let full = ref false in
  Support.Argparse.iter_options flags (function
    | #common_option as o -> Common_options.process_common_option options o
    | `ShowFullDiff -> full := true
  );
  match args with
  | [name] -> (
      match Zeroinstall.Apps.lookup_app options.config name with
      | None -> raise_safe "No such application '%s'" name
      | Some app -> show_app_changes options ~full:!full app
  )
  | _ -> raise (Support.Argparse.Usage_error 1)
